lib/TWCManager/TWCMaster.py

```diff
@@ -269,6 +270,26 @@ def getPricing(self):
             self.exportPricingValues[module["name"]] = module["ref"].getExportPrice()
             self.importPricingValues[module["name"]] = module["ref"].getImportPrice()
 
+    def getPricingInAdvanceAvailable(self):
+        for module in self.getModulesByType("Pricing"):
+            if module["ref"].getPricingInAdvanceAvailable():
+               return True
+        return False
+
+    def getCheaperDayChargeTime(self,dayName):
+        return self.settings["Schedule"][dayName]["cheaper"]
+
+    def getActualHDayChargeTime(self,dayName):
+        return self.settings["Schedule"][dayName]["actualH"]
+
+    def getCheapestStartHour(self,numHours,ini,end):
+        # We take the latest modul data
+        cheapestStartHour = ini
+        for module in self.getModulesByType("Pricing"):
+             cheapestStartHour = module["ref"].getCheapestStartHour(numHours,ini,end)
+
```

Comment:
Just a few thoughts about this:

- numHours should just be end - start
- It's always possible that we could have more than one Pricing module running. The reason might be different data being provided by different modules. It doesn't matter too much but my thoughts are that the logic should be:

```
        for module in self.getModulesByType("Pricing"):
             moduleCheapestHour = module["ref"].getCheapestStartHour(numHours,ini,end)
             if moduleCheapestHour > cheapestStartHour:
                 cheapestStartHour = moduleCheapestHour
```

- This will also require checking the ability for advance pricing per module. We could always just put a dummy function which returns ini if it doesn't support advanced pricing, but it's better to just check and not query if not imho, it saves having to add useless functions

Reply:
Also, just thinking out loud about the scheduled charging logic here:

   - I assume this is really only useful when not using green energy, though. Even the cheapest energy of the day might be more expensive when drawn from the grid rather than green energy. 
   - I think there's somewhat of a hole in the scheduled charging logic, in that we basically look for the cheapest start hour to start a charge for a given duration, but it then may carry over a more expensive period after that, and we don't really protect against charging through a more expensive period overall, we just pick a cheap hour. I worry that people might think they are better off but end up worse off, eg it might trigger a charge to start at 1PM which is peak for renewables but then continue through to 7PM which is peak for grid pricing.
   - It would almost be better to say - how many hours during a 24 hour period do you want to charge your vehicle for, and then we choose the x cheapest hours over that day and start/stop per hour. The fault in that logic is that it assumes a vehicle is connected 24/7, and if it isn't, advanced pricing is not of much use to us as we have no dynamic data to respond to.

I'd be interested in feedback from those with power pricing APIs to hear how they'd want to use the functionality.